bin/propolis-server/src/lib/serial/backend.rs

```diff
@@ -0,0 +1,397 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A backend that provides external clients with access to a Propolis character
+//! device and tracks the history of bytes the guest has written to that device.
+
+use std::{
+    collections::BTreeMap,
+    num::NonZeroUsize,
+    sync::{Arc, Mutex},
+};
+
+use propolis::chardev::{
+    pollers::{SinkBuffer, SourceBuffer},
+    Sink, Source,
+};
+use tokio::{
+    select,
+    sync::{mpsc, oneshot},
+};
+
+use super::history_buffer::{HistoryBuffer, SerialHistoryOffset};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+struct ClientId(u64);
+
+/// A client's rights when accessing a character backend.
+#[derive(Clone, Copy)]
+pub(super) enum Permissions {
+    /// The client may both read and write to the backend.
+    ReadWrite,
+
+    /// The client may only read from the backend.
+    ReadOnly,
+}
+
+impl Permissions {
+    pub(super) fn is_writable(&self) -> bool {
+        matches!(self, Permissions::ReadWrite)
+    }
+}
+
+/// Determines what happens when the backend is unable to send a guest byte back
+/// to a client because the client's channel is full.
+#[derive(Clone, Copy)]
+pub(super) enum FullReadChannelDiscipline {
+    /// The backend should block until it can send to this client.
+    Block,
+
+    /// The backend should close this client's connection.
+    Close,
+}
+
+/// An individual client of a character backend.
+struct Client {
+    /// Bytes read from the character device should be sent to the client on
+    /// this channel.
+    tx: mpsc::Sender<u8>,
+
+    /// Determines what happens when the backend wants to send a byte and
+    /// [`Self::tx`] is full.
+    read_discipline: FullReadChannelDiscipline,
+}
+
+/// A handle held by a client that represents its connection to the backend.
+pub(super) struct ClientHandle {
+    /// The client's ID.
+    id: ClientId,
+
+    /// A reference to the backend to which the client is connected.
+    backend: Arc<ConsoleBackend>,
+
+    /// The client's backend access rights.
+    permissions: Permissions,
+}
```

Comment:
considered nitpicky, take it or leave it: personally, since the `ClientHandle` type represents the external API for consumers of this module, while the `Client` type is an internal-only implementation detail, I might've given the more consicse name to the interface type --- perhaps `Client` and `ClientState`, or something?

this isn't a big deal, though; if renaming it is too much of a pain, feel free to ignore me.